python/tests/test_divmat.py

```diff
@@ -0,0 +1,992 @@
+    lambd[0] = lambd[n]
+    pi[LAMBDA] = 0
+    beta[LAMBDA] = 0
+    alpha[LAMBDA] = 0
+    p = child[LAMBDA]
+    while p != LAMBDA:
+        while True:
+            a = alpha[parent[p]] | (beta[p] & -beta[p])
+            alpha[p] = a
+            if child[p] != LAMBDA:
+                p = child[p]
+            else:
+                break
+        while True:
+            if sib[p] != LAMBDA:
+                p = sib[p]
+                break
+            else:
+                p = parent[p]
+                if p == LAMBDA:
+                    break
+
+    return lambd, pi, tau, beta, alpha
+
+
+def _sv_mrca(x, y, lambd, pi, tau, beta, alpha):
+    if beta[x] <= beta[y]:
+        h = lambd[beta[y] & -beta[x]]
+    else:
+        h = lambd[beta[x] & -beta[y]]
+    k = alpha[x] & alpha[y] & -(1 << h)
+    h = lambd[k & -k]
+    j = ((beta[x] >> h) | 1) << h
+    if j == beta[x]:
+        xhat = x
+    else:
+        ell = lambd[alpha[x] & ((1 << h) - 1)]
+        xhat = tau[((beta[x] >> ell) | 1) << ell]
+    if j == beta[y]:
+        yhat = y
+    else:
+        ell = lambd[alpha[y] & ((1 << h) - 1)]
+        yhat = tau[((beta[y] >> ell) | 1) << ell]
+    if pi[xhat] <= pi[yhat]:
+        z = xhat
+    else:
+        z = yhat
+    return z
+
+
+def sv_mrca(x, y, lambd, pi, tau, beta, alpha):
+    # Convert to 1-based indexes
+    return _sv_mrca(x + 1, y + 1, lambd, pi, tau, beta, alpha) - 1
+
+
+def local_root(tree, u):
+    while tree.parent(u) != tskit.NULL:
+        u = tree.parent(u)
+    return u
+
+
+def divergence_matrix(ts, windows=None, samples=None):
+    windows_specified = windows is not None
+    windows = [0, ts.sequence_length] if windows is None else windows
+    num_windows = len(windows) - 1
+    samples = ts.samples() if samples is None else samples
+
+    n = len(samples)
+    D = np.zeros((num_windows, n, n))
+    tree = tskit.Tree(ts)
+    for i in range(num_windows):
+        left = windows[i]
+        right = windows[i + 1]
+        # print(f"WINDOW {i} [{left}, {right})")
+        tree.seek(left)
+        # Iterate over the trees in this window
+        while tree.interval.left < right and tree.index != -1:
+            span_left = max(tree.interval.left, left)
+            span_right = min(tree.interval.right, right)
+            span = span_right - span_left
+            # print(f"\ttree {tree.interval} [{span_left}, {span_right})")
+            tables = sv_tables_init(tree.parent_array)
+            for j in range(n):
+                u = samples[j]
+                for k in range(j + 1, n):
+                    v = samples[k]
+                    w = sv_mrca(u, v, *tables)
+                    assert w == tree.mrca(u, v)
+                    if w != tskit.NULL:
+                        tu = ts.nodes_time[w] - ts.nodes_time[u]
+                        tv = ts.nodes_time[w] - ts.nodes_time[v]
+                    else:
+                        # Question for @petralharp - is this the right definition?
```

Comment:
I like this definition? Advantages I can think of are (a) it matches what we'd get from the `site` stat, and (b) if you divide the tree into slices by time and compute the contribution to divergence in each time slice, they add to the whole thing?